### PR TITLE
Mesh edit

### PR DESCRIPTION
**Pull request by AlemSnyder**
Adds functions such that if the terrain in the world is edited, then this will update the mesh that is saved on the gpu.

### CODE REVIEW

**Review by AlemSnyder — COMMENTED** [review]
This is a surprisingly small change.

File: src/entity/static_mesh.hpp [hunk 1]
```diff
@@ -21,6 +21,7 @@ class StaticMesh : public MeshLoader::MultiComplexMesh {
     GLuint transforms_buffer_;
     uint32_t num_vertices_;
     uint32_t num_models_;
+    bool do_render_;
 
  public:
     inline StaticMesh(const StaticMesh& obj) {
```
Comment (egelja) [inline src/entity/static_mesh.hpp:27]: Intiializer list

File: src/entity/static_mesh.hpp [hunk 2]
```diff
@@ -21,6 +21,7 @@ class StaticMesh : public MeshLoader::MultiComplexMesh {
     GLuint transforms_buffer_;
     uint32_t num_vertices_;
     uint32_t num_models_;
+    bool do_render_;
```
Comment (egelja) [inline src/entity/static_mesh.hpp:24]:
```suggestion
    GLuint transforms_buffer_;
    uint32_t num_vertices_;
    uint32_t num_models_ : 31;
    bool do_render_ : 1;
```

Comment (egelja) [reply src/entity/static_mesh.hpp:24]: You can save 32 bytes like this

File: src/gui/main_gui.cpp [hunk 3]
```diff
@@ -204,12 +197,15 @@ GUITest(World world) {
     MR.set_light_direction(light_direction);
     MR.set_depth_projection_matrix(depth_projection_matrix);
 
-    for (auto& m : chunk_meshes) {
-        MR.add_mesh(std::make_shared<terrain::TerrainMesh>(m));
+    for (auto m : chunks_mesh) {
+        MR.add_mesh(m);
```
Comment (egelja) [inline src/gui/main_gui.cpp:201]: Drop curlies
Also why are you iterating by copy

File: src/gui/main_gui.cpp [hunk 4]
```diff
@@ -204,12 +197,15 @@ GUITest(World world) {
...
     }
     MR.add_mesh(std::make_shared<terrain::StaticMesh>(treesMesh));
     MR.set_depth_texture(SM.get_depth_texture());
 
+    unsigned int frame_id = 0;
+    bool do_set = false;
```
Comment (egelja) [inline src/gui/main_gui.cpp:207]: This needs a better name

And its also never set to true

File: src/world.cpp [hunk 5]
```diff
@@ -92,7 +98,12 @@ World::World(
         x_tiles, y_tiles, macro_tile_size, height, 5, materials, biome_data["Biome_1"],
         get_grass_grad_data(materials_json),
         materials_json["Dirt"]["Gradient"]["midpoint"].asInt()
-    ) {}
+    ) {
+    chunks_mesh.resize(terrain_main.get_chunks().size());
+    for (size_t i = 0; i < chunks_mesh.size(); i++) {
+        chunks_mesh[i] = std::make_shared<terrain::TerrainMesh>();
+    }
```
Comment (egelja) [inline src/world.cpp:105]: 🚨Code duplication🚨

File: src/world.cpp [hunk 6]
```diff
@@ -102,8 +113,15 @@ World::World(
         macro_tile_size, height, 5, tile_type, materials, biome_data["Biome_1"],
         get_grass_grad_data(materials_json),
         materials_json["Dirt"]["Gradient"]["midpoint"].asInt()
-    ) {}
+    ) {
+    // on initialization world reserves the space it would need for shared pointers
+    chunks_mesh.resize(terrain_main.get_chunks().size());
+    for (size_t i = 0; i < chunks_mesh.size(); i++) {
+        chunks_mesh[i] = std::make_shared<terrain::TerrainMesh>();
+    }
```
Comment (egelja) [inline src/world.cpp:121]: 🚨Code duplication🚨

File: src/world.hpp [hunk 7]
```diff
@@ -50,10 +51,28 @@ class World {
...
+
+    const terrain::Terrain& get_terrain_main() const {
+        return terrain_main;
+    }
+
+    terrain::Terrain& get_terrain_main() {
+        return terrain_main;
+    }
+
+    const std::vector<std::shared_ptr<terrain::TerrainMesh>> get_chunks_mesh() const{
```
Comment (egelja) [inline src/world.hpp:72]: Return a reference

File: src/entity/terrain_mesh.cpp [hunk 8]
```diff
@@ -9,12 +9,28 @@
...
+    GLuint buffers[4] = {
+        vertex_buffer_, color_buffer_, normal_buffer_, element_buffer_};
+    glDeleteBuffers(4, buffers);
+
+    // if indices are none so if there is no vertices that would be sent to the graphics
+    // card
+    //     then there is no reason to create a buffer
+    // create a bool do_render, set to false when
+    num_vertices_ = mesh.get_indices().size();
+    do_render_ = (num_vertices_ != 0);
```
[the remark this answers is not in the document]
Comment (AlemSnyder) [reply src/entity/terrain_mesh.cpp:27]: no

File: src/gui/main_gui.cpp (hunk 4 above)
Comment (AlemSnyder) [reply src/gui/main_gui.cpp:207]: used to test, this is manualy set to true.

File: src/world.cpp [hunk 9]
```diff
@@ -102,8 +113,15 @@ World::World(
...
+}
 
+// ! deprecated
```
[the remark this answers is not in the document]
Comment (AlemSnyder) [reply src/world.cpp:124]: this gets removed so it dosen't matter

File: src/world.cpp (hunk 6 above)
Comment (AlemSnyder) [reply src/world.cpp]: it's three lines of code

Comment (egelja) [reply src/world.cpp]: and if you change it in one place you have to remember to change it in three.

**Comment by AlemSnyder** [comment]
All comments resolved

File: src/world.hpp [hunk 10]
```diff
@@ -50,10 +51,30 @@ class World {
...
+    get_terrain_main() const {
+        return terrain_main;
+    }
+
+    terrain::Terrain&
+    get_terrain_main() {
+        return terrain_main;
+    }
+
+    const std::vector<std::shared_ptr<terrain::TerrainMesh>>&
```
Comment (egelja) [inline src/world.hpp:73]:
```suggestion
    const auto&
```

You don't need to repeat the type, its given on L57 already.

File: src/world.hpp [hunk 11]
```diff
@@ -126,4 +148,47 @@ class World {
...
+    }
+
+    inline void
+    qb_save(const std::string& path) const {
+        terrain_main.qb_save(path);
+    }
+
+ private:
+    inline void
+    initialize_chunks_mesh() {
```
Comment (egelja) [inline src/world.hpp:189]:
```suggestion
    initialize_chunks_mesh_() {
```

**Review by egelja — CHANGES_REQUESTED** [review]
Use `auto` more and avoid repeating types

File: src/entity/terrain_mesh.cpp [hunk 12]
```diff
@@ -46,6 +62,20 @@ terrain::TerrainMesh::init(const entity::Mesh& mesh) {
         GL_ELEMENT_ARRAY_BUFFER, mesh.get_indices().size() * sizeof(unsigned short),
         mesh.get_indices().data(), GL_STATIC_DRAW
     );
+}
 
-    num_vertices_ = mesh.get_indices().size();
-}
\ No newline at end of file
+terrain::TerrainMesh&
+terrain::TerrainMesh::operator=(TerrainMesh&& other) {
```
[the remark this answers is not in the document]
Comment (AlemSnyder) [reply src/entity/terrain_mesh.cpp:68]: remember that these uints are actualy pointers. Only one object can have accesses to them.

File: src/world.hpp (hunk 10 above)
Comment (AlemSnyder) [reply src/world.hpp:73]: I don't like this

File: src/entity/static_mesh.hpp [hunk 13]
```diff
@@ -17,24 +17,19 @@ class StaticMesh : public MeshLoader::MultiComplexMesh {
...
     GLuint normal_buffer_;
     GLuint element_buffer_;
-    GLuint color_texture_;
     GLuint transforms_buffer_;
+    GLuint color_texture_;
     uint32_t num_vertices_;
-    uint32_t num_models_;
-    bool do_render_;
+    uint32_t num_models_ : 31;
+    bool do_render_ : 1;
```
[the remark this answers is not in the document]
Comment (AlemSnyder) [reply src/entity/static_mesh.hpp]: neither are boing to higher than 3 million, so I don't think it maters.
num models should be much smaller.

File: src/entity/terrain_mesh.cpp (hunk 12 above)
[the remark this answers is not in the document]
Comment (egelja) [reply src/entity/terrain_mesh.cpp]: Yes that is why you do this:

```c++
terrain::TerrainMesh::operator=(TerrainMesh&& other) = default;

terrain::TerrainMesh::operator=(const TerrainMesh& other) = delete;
```

**Review by egelja — CHANGES_REQUESTED** [review]

File: src/world.cpp (hunk 9 above)
[the remark this answers is not in the document]
Comment (egelja) [reply src/world.cpp:114]: then remove it

File: src/world.cpp [hunk 14]
```diff
@@ -120,3 +128,59 @@ World::get_mesh_greedy() const {
...
+}
+
+void
+World::update_single_mesh(TerrainDim3 tile_sop) {
+    if (terrain_main.in_range(tile_sop.x + 1, tile_sop.y, tile_sop.z)) {
+        uint16_t chunk_pos =
+            terrain_main.get_chunk_from_tile(tile_sop.x + 1, tile_sop.y, tile_sop.z);
+        update_single_mesh(chunk_pos);
+    }
+}
```
Comment (egelja) [inline src/world.cpp:157]: define these in a sensical order

```suggestion
void
World::update_single_mesh(uint16_t chunk_pos) {
    const auto& chunks = terrain_main.get_chunks();
    entity::Mesh chunk_mesh = entity::generate_mesh(chunks[chunk_pos]);

    chunk_mesh.change_color_indexing(
        materials, terrain::TerrainColorMapping::get_colors_inverse_map()
    );

    chunks_mesh[chunk_pos]->update(chunk_mesh);
}

void
World::update_single_mesh(TerrainDim3 tile_sop) {
    if (terrain_main.in_range(tile_sop.x + 1, tile_sop.y, tile_sop.z)) {
        uint16_t chunk_pos =
            terrain_main.get_chunk_from_tile(tile_sop.x + 1, tile_sop.y, tile_sop.z);
        update_single_mesh(chunk_pos);
    }
}

void
World::update_all_chunks_mesh() {
    for (size_t i = 0; i < chunks_mesh.size(); i++)
        update_single_mesh(i);
}
```

File: src/world.cpp [hunk 15]
```diff
@@ -120,3 +128,59 @@ World::get_mesh_greedy() const {
...
+
+void
+World::set_tile(uint16_t pos, const terrain::Material* mat, uint8_t color_id) {
```
Comment (egelja) [inline src/world.cpp:160]: TYPEDEFS for `pos` and `color_id`

File: src/world.cpp [hunk 16]
```diff
@@ -120,3 +128,59 @@ World::get_mesh_greedy() const {
...
+    terrain_main.get_tile(pos)->set_material(mat, color_id);
+
+    TerrainDim3 tile_sop = terrain_main.sop(pos);
+    update_single_mesh(tile_sop);
+
+    // do some math:
+    // if the tile is on the edge of a chunk then both chunks must be updated.
+    uint8_t edge_case = tile_sop.x % terrain::Chunk::SIZE;
```
Comment (egelja) [inline src/world.cpp:168]: use the same type as `pos` (after you typedef it)

File: src/world.cpp [hunk 17]
```diff
@@ -120,3 +128,59 @@ World::get_mesh_greedy() const {
...
+        update_single_mesh({tile_sop.x, tile_sop.y - 1, tile_sop.z});
+    } else if (edge_case == terrain::Chunk::SIZE - 1) {
+        update_single_mesh({tile_sop.x, tile_sop.y + 1, tile_sop.z});
+    }
+    edge_case = tile_sop.z % terrain::Chunk::SIZE;
+    if (edge_case == 0) {
+        update_single_mesh({tile_sop.x, tile_sop.y, tile_sop.z - 1});
+    } else if (edge_case == terrain::Chunk::SIZE - 1) {
+        update_single_mesh({tile_sop.x, tile_sop.y, tile_sop.z + 1});
+    }
```
Comment (egelja) [inline src/world.cpp:185]: spacing and formatting

```suggestion
    if (edge_case == 0)
        update_single_mesh({tile_sop.x - 1, tile_sop.y, tile_sop.z});
    else if (edge_case == terrain::Chunk::SIZE - 1)
        update_single_mesh({tile_sop.x + 1, tile_sop.y, tile_sop.z});
    
    edge_case = tile_sop.y % terrain::Chunk::SIZE;
    if (edge_case == 0)
        update_single_mesh({tile_sop.x, tile_sop.y - 1, tile_sop.z});
    else if (edge_case == terrain::Chunk::SIZE - 1)
        update_single_mesh({tile_sop.x, tile_sop.y + 1, tile_sop.z});
    
    edge_case = tile_sop.z % terrain::Chunk::SIZE;
    if (edge_case == 0)
        update_single_mesh({tile_sop.x, tile_sop.y, tile_sop.z - 1});
    else if (edge_case == terrain::Chunk::SIZE - 1)
        update_single_mesh({tile_sop.x, tile_sop.y, tile_sop.z + 1});
```

File: src/world.hpp [hunk 18]
```diff
@@ -50,10 +51,30 @@ class World {
     // materials that exist
     std::map<MaterialId, const terrain::Material> materials;
 
- public:
     // terrain in the world
     terrain::Terrain terrain_main;
```
Comment (egelja) [inline src/world.hpp:55]:
```suggestion
    terrain::Terrain terrain_main_;
```

File: src/world.hpp [hunk 19]
```diff
@@ -50,10 +51,30 @@ class World {
...
 
+    // TerrainMesh for each chunk in terrain
+    std::vector<std::shared_ptr<terrain::TerrainMesh>> chunks_mesh;
```
Comment (egelja) [inline src/world.hpp:58]:
```suggestion
    std::vector<std::shared_ptr<terrain::TerrainMesh>> chunks_mesh_;
```

File: src/world.hpp [hunk 20]
```diff
@@ -50,10 +51,30 @@ class World {
...
+
+    // chunks_mesh like attorneys general
+
+ public:
+    const terrain::Terrain&
```
Comment (egelja) [inline src/world.hpp:63]:
```suggestion
    const auto&
```

File: src/world.hpp [hunk 21]
```diff
@@ -50,10 +51,30 @@ class World {
...
+    get_terrain_main() const {
+        return terrain_main;
+    }
+
+    terrain::Terrain&
```
Comment (egelja) [inline src/world.hpp:68]:
```suggestion
    auto&
```

File: src/world.hpp [hunk 22]
```diff
@@ -75,8 +96,8 @@ class World {
      */
     World(const Json::Value& materials_json, const Json::Value& biome_data, int type);
     World(
-        const Json::Value& materials_json, const Json::Value& biome_data, uint32_t x_tiles,
-        uint32_t y_tiles
+        const Json::Value& materials_json, const Json::Value& biome_data,
+        uint32_t x_tiles, uint32_t y_tiles
     );
```
Comment (egelja) [inline src/world.hpp:101]: typedefs

File: src/world.hpp [hunk 23]
```diff
@@ -126,4 +148,47 @@ class World {
...
+    void stamp_tile_region(
+        int x_start, int y_start, int z_start, int x_end, int y_end, int z_end,
+        const terrain::Material* mat, std::set<std::pair<int, int>> elements_can_stamp,
+        uint8_t color_id
+    );
+
+    void stamp_tile_region(
+        int x_start, int y_start, int z_start, int x_end, int y_end, int z_end,
+        const terrain::Material* mat, uint8_t color_id
+    );
```
Comment (egelja) [inline src/world.hpp:175]: TYPEDEFS

File: src/world.hpp [hunk 24]
```diff
@@ -126,4 +148,47 @@ class World {
...
+
+    inline void
+    qb_save_debug(const std::string& path) {
```
Comment (egelja) [inline src/world.hpp:178]:
```suggestion
    qb_save_debug(const std::string& path) const {
```

**Review by egelja — CHANGES_REQUESTED** [review]

File: src/world.cpp (hunk 9 above)
[the remark this answers is not in the document]
Comment (AlemSnyder) [reply src/world.cpp:114]: then you should have only reviewed the most recent pull request, they are all merged together.

File: src/world.hpp (hunk 24 above)
Comment (AlemSnyder) [reply src/world.hpp:178]: it's not constant. it changes the data (ya I know its not grate, but i didn't want to copy all of terrain)

File: src/gui/main_gui.cpp [hunk 25]
```diff
@@ -135,21 +131,18 @@ GUITest(World world) {
     glBindVertexArray(VertexArrayID);
 
     //  The mesh of the terrain
-    std::vector<terrain::TerrainMesh> chunk_meshes;
-    chunk_meshes.resize(mesh.size());
-    for (size_t i = 0; i < chunk_meshes.size(); i++) {
-        chunk_meshes[i].init(mesh[i]);
-    }
+    world.update_all_chunks_mesh();
```
[the remark this answers is not in the document]
Comment (AlemSnyder) [reply src/gui/main_gui.cpp:134]: no

File: src/entity/terrain_mesh.cpp [hunk 26]
```diff
@@ -46,6 +68,22 @@ terrain::TerrainMesh::init(const entity::Mesh& mesh) {
...
+    vertex_buffer_ = std::exchange(other.vertex_buffer_, 0);
+    color_buffer_ = std::exchange(other.color_buffer_, 0);
+    normal_buffer_ = std::exchange(other.normal_buffer_, 0);
+    element_buffer_ = std::exchange(other.element_buffer_, 0);
+    color_texture_ = std::exchange(other.color_texture_, 0);
+    num_vertices_ = std::exchange(other.num_vertices_, 0);
+    do_render_ = std::exchange(other.do_render_, false);
+
+    return *this;
+}
```
[the remark this answers is not in the document]
Comment (AlemSnyder) [reply src/entity/terrain_mesh.cpp]: I feel like I've gone over this. Eeach of these is like a pointer. They cannot sipley be coppied. There must be singular ownership.

File: src/world.hpp (hunk 20 above)
Comment (AlemSnyder) [reply src/world.hpp:63]: no

File: src/world.hpp (hunk 21 above)
Comment (AlemSnyder) [reply src/world.hpp:68]: no

File: src/terrain/path/tile_iterators.hpp [hunk 27]
```diff
@@ -58,12 +58,16 @@ class AdjacentIterator {
     int operator++();
     int operator++(int);
 
-    inline bool end() { return dpos_ > 26U; }
+    inline bool
+    end() {
+        return dpos_ > 26U;
+    }
```
[the remark this answers is not in the document]
Comment (AlemSnyder) [reply src/terrain/path/tile_iterators.hpp:64]: I think end is better. end is used for pointers, so I think it is clearer this way.

File: src/util/bits.hpp
[the remark this answers is not in the document]
Comment (AlemSnyder) [reply src/util/bits.hpp:1]: lol too modern for debian

File: src/entity/static_mesh.cpp
[the remark this answers is not in the document]
Comment (AlemSnyder) [reply src/entity/static_mesh.cpp:1]: I have committed many a cpp crime

[the remark this answers is not in the document]
Comment (AlemSnyder) [reply src/entity/static_mesh.cpp:1]:
```error: call of overloaded ‘add_mesh(std::shared_ptr<terrain::StaticMesh>)’ is ambiguous
  193 |     SM.add_mesh(std::make_shared<terrain::StaticMesh>(treesMesh));```
So this is the issue

[the remark this answers is not in the document]
Comment (AlemSnyder) [reply src/entity/static_mesh.cpp:1]: My computer didn't want to copy the error correctly, but the issue is that TerrainMesh, and StaticMesh need to be differentiated when they are pointers/virtual classes. I cold have them both inherit from some third thing, but that seems worse than then this code duplication.

File: src/entity/terrain_mesh.cpp (hunk 26 above)
[the remark this answers is not in the document]
Comment (egelja) [reply src/entity/terrain_mesh.cpp]: This is the **move** ctor. Not the **copy** ctor, which should be deleted.

File: src/world.hpp (hunk 21 above)
Comment (egelja) [reply src/world.hpp:68]: Yes this isn’t Java

File: src/world.hpp (hunk 24 above)
Comment (egelja) [reply src/world.hpp:177]: Why does it change the data, when writing to disk?

File: src/entity/static_mesh.cpp
[the remark this answers is not in the document]
Comment (egelja) [reply src/entity/static_mesh.cpp:1]: So you have a ptr to a MultiComplexMesh and need to downcast it?

the right way to do this is a pure virtual on the base class and concrete virtual methods on the subclasses. 

The slow way is to do a bunch of `dynamic_cast`s, which use `strcmp` internally.

File: src/world.hpp (hunk 24 above)
Comment (AlemSnyder) [reply src/world.hpp:177]: it sets the color of air aire blocks to a color depending on their chunk and path connected tiles.

Theorietically I could pass a different `get_voxel` function.

File: src/entity/static_mesh.cpp
[the remark this answers is not in the document]
Comment (AlemSnyder) [reply src/entity/static_mesh.cpp:1]: Currently it is like this, and woks.

```
SingleComplexMesh       MultiComplexMesh  <- pure virtual
     |                        |
     V                        V
TerrainMesh             StaticMesh        <- concreate virtual```

You want to reduce code duplication. I assume with inheratance.

```
SingleComplexMesh <- pure virtual
     |
     V
TerrainMesh             MultiComplexMesh <- pure virtual
        |              /
        V             V
          StaticMesh
```

**Comment by AlemSnyder** [comment]
I messed up logging in opengl. This leads to segfaults...

The issue is fixed on Graphics/imgui. That pull request also includes mesh edit, and sky background 2.

I don't want to deal with two extra prs, so I will close these two, and merge everything into imgui.

**Pull request closed without merging**